src/services/eventManager.js

```diff
@@ -0,0 +1,96 @@
+(function () {
+    angular.module('ui-leaflet')
+        .factory('eventManager', [
+            function () {
+                var EventManager = function() {
+                    this.listeners = {};
+                };
+
+                EventManager.prototype = {
+                    addEventListener: function (type, callback, scope) {
+                        var args = [];
+                        var numOfArgs = arguments.length;
+                        for (var i = 0; i < numOfArgs; i++) {
+                            args.push(arguments[i]);
+                        }
+                        args = args.length > 3 ? args.splice(3, args.length - 1) : [];
+                        if (typeof this.listeners[type] !== "undefined") {
+                            this.listeners[type].push({scope: scope, callback: callback, args: args});
+                        } else {
+                            this.listeners[type] = [{scope: scope, callback: callback, args: args}];
+                        }
+                    },
+                    removeEventListener: function (type, callback, scope) {
+                        if (typeof this.listeners[type] !== "undefined") {
+                            var numOfCallbacks = this.listeners[type].length;
+                            var newArray = [];
+                            for (var i = 0; i < numOfCallbacks; i++) {
+                                var listener = this.listeners[type][i];
+                                if (listener.scope === scope && listener.callback === callback) {
+
+                                } else {
+                                    newArray.push(listener);
+                                }
+                            }
+                            this.listeners[type] = newArray;
+                        }
+                    },
+                    hasEventListener: function (type, callback, scope) {
+                        if (typeof this.listeners[type] !== "undefined") {
+                            var numOfCallbacks = this.listeners[type].length;
+                            if (callback === undefined && scope === undefined) {
+                                return numOfCallbacks > 0;
```

Comment:
Instead of all the for loops we should be using our own iterator. https://github.com/angular-ui/ui-leaflet/blob/master/src/services/leafletIterators.js#L156